[Fleet] Sync integration status UI

juliaElastic opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/217154

To test locally:
 - Follow this guide to set up 2 clusters locally: https://github.com/elastic/kibana/blob/main/x-pack/platform/plugins/shared/fleet/dev_docs/local_setup/remote_clusters_ccr.md
 - Install a few integrations on the main cluster and create a few custom component templates and ingest pipelines
 - Go to Fleet Settings, check the Sync status in the Output table

### Screenshots

Output table

<img width="1096" alt="image" src="https://github.com/user-attachments/assets/047b516a-b32a-4827-a943-de1119d45dbe" />

Sync integrations status flyout 
- Added mock response to show the different UI states
- Added `Close` button instead of `Cancel` and `Done` because there is no action to take on the flyout, it seemed unnecessary

<img width="598" alt="image" src="https://github.com/user-attachments/assets/7cc70721-a765-488b-8191-a8a0aaefe4a1" />

 Tooltips

<img width="532" alt="image" src="https://github.com/user-attachments/assets/387cdf84-e807-4287-8802-4a512c756a3a" />
<img width="343" alt="image" src="https://github.com/user-attachments/assets/3c947361-5de8-40c2-bab8-a73e6321e9a8" />

Top level error without any integrations

<img width="605" alt="image" src="https://github.com/user-attachments/assets/78d9a79a-fac5-4af2-9745-46dbdbe956a2" />




### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

juliaElastic committed:
sync integration status UI

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

juliaElastic committed:
use memo

juliaElastic committed:
avoid unnecessary package info requests

juliaElastic commented on x-pack/platform/plugins/shared/fleet/public/applications/fleet/sections/settings/components/outputs_table/integration_status.tsx:
copied and adjusted from https://github.com/elastic/kibana/blob/main/x-pack/platform/plugins/shared/fleet/public/applications/fleet/sections/agents/agent_details_page/components/agent_details/agent_details_integration.tsx

juliaElastic committed:
show loading spinner for synchronizing custom assets

juliaElastic committed:
fix test

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/293598)
* Commit: ef12634e6bf13a769f3b598c513a8d90a1e680b4

### Failed CI Steps
* [Jest Tests #11](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e00-a74d-4a75-b308-223e9dedf3da)
* [FTR Configs #21](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e00-a769-4ad3-97c7-2218e4dbe7be)
* [FTR Configs #48](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e00-a77e-4775-b90c-4e4bbd4be53f)
* [FTR Configs #48](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e2a-0404-4f77-b498-7b6e23668246)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e00-a77e-4775-b90c-4e4bbd4be53f) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/293598/jobs/01963e00-a77e-4775-b90c-4e4bbd4be53f/artifacts/01963e29-ba6f-4c8f-b372-ecf600f2d93b) FTR Configs #<span></span>48 / @<span></span>ess @<span></span>serverless SecuritySolution Explore Overview Overview Host With auditbeat Make sure that we get OverviewHost data
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e2a-0404-4f77-b498-7b6e23668246) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/293598/jobs/01963e2a-0404-4f77-b498-7b6e23668246/artifacts/01963e31-06dd-4b23-a9f0-b9917ffb6bfb) FTR Configs #<span></span>48 / @<span></span>ess @<span></span>serverless SecuritySolution Explore Overview Overview Host With auditbeat Make sure that we get OverviewHost data
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e00-a769-4ad3-97c7-2218e4dbe7be) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/293598/jobs/01963e00-a769-4ad3-97c7-2218e4dbe7be/artifacts/01963e27-56e4-4675-bd32-93d39199415c) FTR Configs #<span></span>21 / console app misc console behavior keyboard shortcuts open documentation should open documentation when Ctrl+/ is pressed
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e00-a74d-4a75-b308-223e9dedf3da) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/293598/jobs/01963e00-a74d-4a75-b308-223e9dedf3da/artifacts/01963e25-b559-473e-b8cf-bb626c4d588b) Jest Tests #<span></span>11 / getRemoteSyncedIntegrationsInfoByOutputId should not throw if the remote status api has errors in the body
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/293598#01963e00-a74d-4a75-b308-223e9dedf3da) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/293598/jobs/01963e00-a74d-4a75-b308-223e9dedf3da/artifacts/01963e25-b558-42ee-808b-eea27ced622b) Jest Tests #<span></span>11 / getRemoteSyncedIntegrationsInfoByOutputId should throw error if the output returns undefined

cc @juliaElastic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293598","number":293598,"commit":"ef12634e6bf13a769f3b598c513a8d90a1e680b4"}],"number":293598}
buildkite-pr-comment-->

juliaElastic committed:
Merge branch 'main' into sync-integrations-ui

juliaElastic committed:
update error message

juliaElastic committed:
add UI tests

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

juliaElastic commented on x-pack/platform/plugins/shared/fleet/server/tasks/sync_integrations/get_remote_status.ts:
```diff
@@ -63,14 +60,25 @@ export const getRemoteSyncedIntegrationsInfoByOutputId = async (
 
-    return body as GetRemoteSyncedIntegrationsStatusResponse;
+    const errorMessage = body.error
+      ? `GET ${url} failed with status ${body.statusCode}. ${body.message}`
```
while testing with an invalid Kibana API Key, getting a body like this:
```
{
  statusCode: 401,
  error: 'Unauthorized',
  message: '[security_exception\n' +
    '\tRoot causes:\n' +
    '\t\tsecurity_exception: unable to authenticate with provided credentials and anonymous access is not allowed for this request]: unable to authenticate with provided credentials and anonymous access is not allowed for this request'
}
```

juliaElastic committed:
update test

jillguyonnet commented:
Testing this now - just mentioning [this comment](https://github.com/elastic/kibana/pull/217799#issuecomment-2804478766) for extra visibility which I added to https://github.com/elastic/kibana/issues/217683. If it gets addressed there, an invalid remote Kibana URL should surface a more helpful error than currently:

<img width="956" alt="Screenshot 2025-04-16 at 15 47 26" src="https://github.com/user-attachments/assets/4bb38cac-53bc-41d9-8b85-685a2e8368e2" />

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/293750)
* Commit: 3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3

### Failed CI Steps
* [Jest Integration Tests #7](https://buildkite.com/elastic/kibana-pull-request/builds/293750#01963ee2-38df-4f78-bdd4-ee5f30be12ad)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/19feb59109b2cbc53667042e04ceba4f4e92c901) | [after](https://github.com/elastic/kibana/commit/3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3) | diff |
| --- | --- | --- | --- |
| `fleet` | 1205 | 1209 | +4 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/19feb59109b2cbc53667042e04ceba4f4e92c901) | [after](https://github.com/elastic/kibana/commit/3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3) | diff |
| --- | --- | --- | --- |
| `fleet` | 1.7MB | 1.7MB | +8.2KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/19feb59109b2cbc53667042e04ceba4f4e92c901) | [after](https://github.com/elastic/kibana/commit/3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3) | diff |
| --- | --- | --- | --- |
| `fleet` | 163.4KB | 164.1KB | +675.0B |



### History
* :broken_heart: [Build #293598](https://buildkite.com/elastic/kibana-pull-request/builds/293598) failed ef12634e6bf13a769f3b598c513a8d90a1e680b4

cc @juliaElastic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293750","number":293750,"commit":"3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293598","number":293598,"commit":"ef12634e6bf13a769f3b598c513a8d90a1e680b4"}],"number":293750}
buildkite-pr-comment-->

juliaElastic committed:
catch invalid json error

juliaElastic commented:
> Testing this now - just mentioning [this comment](https://github.com/elastic/kibana/pull/217799#issuecomment-2804478766) for extra visibility which I added to #217683. If it gets addressed there, an invalid remote Kibana URL should surface a more helpful error than currently:

Fixed the invalid Kibana URL scenario by catching the error in `res.json()` and returning an error message:
```
{
    "integrations": [],
    "error": "GET http://localhost:5701/julia/invalid/api/fleet/remote_synced_integrations/status failed with status 404. Not Found"
}
```

jillguyonnet reviewed (approved):
LGTM 🎉

juliaElastic committed:
hide sync output table column when feature flag is disabled

juliaElastic commented on x-pack/platform/plugins/shared/fleet/public/applications/fleet/sections/settings/components/outputs_table/index.tsx:
```diff
@@ -117,6 +121,16 @@ export const OutputsTable: React.FunctionComponent<OutputsTableProps> = ({
           defaultMessage: 'Status',
         }),
       },
+      ...(enableSyncIntegrationsOnRemote
```
hiding the column if feature flag is off

<img width="1185" alt="image" src="https://github.com/user-attachments/assets/746806ca-5f9c-43b6-b655-6936262dada9" />

juliaElastic committed:
Merge branch 'main' into sync-integrations-ui

juliaElastic committed:
return error from remote API

juliaElastic commented on x-pack/platform/plugins/shared/fleet/server/tasks/sync_integrations/get_remote_status.ts:
```diff
@@ -63,14 +60,30 @@ export const getRemoteSyncedIntegrationsInfoByOutputId = async (
+    return {
+      integrations: body?.integrations ?? [],
+      custom_assets: body?.custom_assets,
+      error: errorMessage ?? body?.error,
```
Returning error from remote API, e.g. when follower index is paused
<img width="607" alt="image" src="https://github.com/user-attachments/assets/ab6fea8c-15f9-4f3e-9ef9-0f4856c0c685" />

juliaElastic committed:
filter out integrations not_installed in status API

juliaElastic committed:
return error.message

juliaElastic committed:
fix installing status

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/294165)
* Commit: e6be997e33cc01979eae2a82d4758bef49ae732a

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/41ccdb70da28959a9025dab4dfbf36c58228eb67) | [after](https://github.com/elastic/kibana/commit/e6be997e33cc01979eae2a82d4758bef49ae732a) | diff |
| --- | --- | --- | --- |
| `fleet` | 1208 | 1212 | +4 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/41ccdb70da28959a9025dab4dfbf36c58228eb67) | [after](https://github.com/elastic/kibana/commit/e6be997e33cc01979eae2a82d4758bef49ae732a) | diff |
| --- | --- | --- | --- |
| `fleet` | 1.7MB | 1.7MB | +8.3KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/41ccdb70da28959a9025dab4dfbf36c58228eb67) | [after](https://github.com/elastic/kibana/commit/e6be997e33cc01979eae2a82d4758bef49ae732a) | diff |
| --- | --- | --- | --- |
| `fleet` | 163.4KB | 164.1KB | +675.0B |



### History
* :green_heart: [Build #294012](https://buildkite.com/elastic/kibana-pull-request/builds/294012) succeeded cd2901c93c9e9b87c04fc94ea3feb065b33844c4
* :green_heart: [Build #293834](https://buildkite.com/elastic/kibana-pull-request/builds/293834) succeeded f441416763622a93fb7e57777950602b495b621f
* :yellow_heart: [Build #293750](https://buildkite.com/elastic/kibana-pull-request/builds/293750) was flaky 3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3
* :broken_heart: [Build #293598](https://buildkite.com/elastic/kibana-pull-request/builds/293598) failed ef12634e6bf13a769f3b598c513a8d90a1e680b4

cc @juliaElastic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/294165","number":294165,"commit":"e6be997e33cc01979eae2a82d4758bef49ae732a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/294012","number":294012,"commit":"cd2901c93c9e9b87c04fc94ea3feb065b33844c4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293834","number":293834,"commit":"f441416763622a93fb7e57777950602b495b621f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293750","number":293750,"commit":"3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293598","number":293598,"commit":"ef12634e6bf13a769f3b598c513a8d90a1e680b4"}],"number":294165}
buildkite-pr-comment-->

juliaElastic committed:
Merge branch 'main' into sync-integrations-ui

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/294753)
* Commit: 3a9d22812625f0ac43796aa8310f53d4861119b1

### Failed CI Steps
* [FTR Configs #7](https://buildkite.com/elastic/kibana-pull-request/builds/294753#01965c73-f852-4206-a56b-e73c2889dacc)
* [FTR Configs #117](https://buildkite.com/elastic/kibana-pull-request/builds/294753#01965c73-f89d-4b1f-bdee-0d1fbfe82f06)
* [FTR Configs #131](https://buildkite.com/elastic/kibana-pull-request/builds/294753#01965c73-f8a6-41c4-a775-88180d132373)
* [Fleet Cypress Tests #1](https://buildkite.com/elastic/kibana-pull-request/builds/294753#01965c58-c069-4de8-a2a7-97e417d859f0)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/294753#01965c73-f852-4206-a56b-e73c2889dacc) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/294753/jobs/01965c73-f852-4206-a56b-e73c2889dacc/artifacts/01965c9c-df51-468a-8dd5-4da105ef9674) FTR Configs #<span></span>7 / serverless search UI - feature flags Serverless Synonyms Overview Synonyms get started Page should not override existing synonyms set unless specified
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/294753#01965c58-c069-4de8-a2a7-97e417d859f0) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/294753/jobs/01965c58-c069-4de8-a2a7-97e417d859f0/artifacts/01965c96-7a96-4b57-833b-745e5d50d10d) Fleet Cypress Tests #<span></span>1 / View agents list Bulk actions should allow to bulk upgrade agents and cancel that upgrade

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/f8354b0770976e2a4ac3f4e11acd6fc7cbbf5b38) | [after](https://github.com/elastic/kibana/commit/3a9d22812625f0ac43796aa8310f53d4861119b1) | diff |
| --- | --- | --- | --- |
| `fleet` | 1186 | 1190 | +4 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/f8354b0770976e2a4ac3f4e11acd6fc7cbbf5b38) | [after](https://github.com/elastic/kibana/commit/3a9d22812625f0ac43796aa8310f53d4861119b1) | diff |
| --- | --- | --- | --- |
| `fleet` | 1.7MB | 1.7MB | +8.3KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/f8354b0770976e2a4ac3f4e11acd6fc7cbbf5b38) | [after](https://github.com/elastic/kibana/commit/3a9d22812625f0ac43796aa8310f53d4861119b1) | diff |
| --- | --- | --- | --- |
| `fleet` | 163.4KB | 164.1KB | +675.0B |



### History
* :green_heart: [Build #294165](https://buildkite.com/elastic/kibana-pull-request/builds/294165) succeeded e6be997e33cc01979eae2a82d4758bef49ae732a
* :green_heart: [Build #294012](https://buildkite.com/elastic/kibana-pull-request/builds/294012) succeeded cd2901c93c9e9b87c04fc94ea3feb065b33844c4
* :green_heart: [Build #293834](https://buildkite.com/elastic/kibana-pull-request/builds/293834) succeeded f441416763622a93fb7e57777950602b495b621f
* :yellow_heart: [Build #293750](https://buildkite.com/elastic/kibana-pull-request/builds/293750) was flaky 3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3
* :broken_heart: [Build #293598](https://buildkite.com/elastic/kibana-pull-request/builds/293598) failed ef12634e6bf13a769f3b598c513a8d90a1e680b4

cc @juliaElastic

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/294753","number":294753,"commit":"3a9d22812625f0ac43796aa8310f53d4861119b1"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/294165","number":294165,"commit":"e6be997e33cc01979eae2a82d4758bef49ae732a"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/294012","number":294012,"commit":"cd2901c93c9e9b87c04fc94ea3feb065b33844c4"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293834","number":293834,"commit":"f441416763622a93fb7e57777950602b495b621f"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293750","number":293750,"commit":"3c3015cb3dbec2b2d543ddce0feda8f55f38b9f3"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/293598","number":293598,"commit":"ef12634e6bf13a769f3b598c513a8d90a1e680b4"}],"number":294753}
buildkite-pr-comment-->
